api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto

```diff
@@ -75,6 +76,15 @@ message HttpProtocolOptions {
     config.core.v3.Http2ProtocolOptions http2_protocol_options = 2;
   }
 
+  // If this is used, the cluster can will use both HTTP/1 and HTTP/2, whichever
+  // protocol is negotiated by ALPN with the upstream.
+  // If the upstream does not support ALPN, it will fail over to HTTP/1.
+  message AlpnHttpConfig {
```

Comment:
We had a conversation about this in the other PR I think, but I'm wondering if this should actually not contain Alpn in the name? Do we want to ultimately use this for happy eyeballs and QUIC -> TCP fallback (and possible connection racing for that as well as v4 vs v6)?

Should we call this `AutoDiscoveredHttpConfig`? And then setup the message to eventually support various future proof config for ALPN/Happy eyeballs/etc.?

Also, for ALPN, can you clarify what ALPN values should be set on the transport and whether anything will be checked? I can't remember where we landed there but I wonder how we can make sure that users don't run into confusing config issues here and also whether we need to allow overriding the ALPN values that are set?

Reply:
Yeah, updated w.r.t. ALPN, thanks.  I think with the name change this works fine for QUIC, as we'd just add HTTP/3 as optional and update the comment w.r.t. trying dual probing and then ALPN for TCP.
I don't think we'd do happy eyeballs here - I think we'd do them in cluster config since they apply to TCP as well.  WDYT?

Reply:
> I don't think we'd do happy eyeballs here - I think we'd do them in cluster config since they apply to TCP as well. WDYT?

Yeah I could go either way on this. I think it's generally a property of the connection pool, and as more of the code is shared between HTTP and TCP maybe this just becomes a common config option that is also shared between them?

Reply:
Sure, but again I think that's orthogonal to this PR?  I could see people wanting auto-ipv4/v6 with explicit htttp config, so I don't think it goes in this new proto.  WDYT?

Reply:
Yes I wasn't suggesting to do it now, I'm just trying to make sure we think through the API that we might need in the future. I think maybe we would embed some common TCP option API in this one and then also allow that option to be used in typed_config for raw TCP protocol?